utils/transaction-utils.js

```diff
@@ -397,107 +398,348 @@ const signAndPublishSlpTransaction = async (
-          `Address to receive swept funds must be included unless balanceOnly flag is true.`
-        );
-      }
+    if (!addressBch || addressBch === "") {
+      throw new Error(`Address to receive swept BCH funds must be included`);
+    }
+    if (!addressSlp || addressSlp === "") {
+      throw new Error(`Address to receive swept SLP funds must be included`);
     }
-    // Generate a keypair from the WIF.
-    const keyPair = SLP.ECPair.fromWIF(wif);
 
-    // Generate the public address associated with the private key.
-    const fromAddr: string = SLP.ECPair.toCashAddress(keyPair);
+    if (tokenId && tokenDecimals == null) {
+      throw new Error("Token decimals required");
+    }
 
-    // Check the BCH balance of that public address.
-    const details = await SLP.Address.details(fromAddr);
-    const balance: number = details.balance;
+    let txid = null;
 
-    // If balance is zero or balanceOnly flag is passed in, exit.
-    if (balance === 0 || balanceOnly) return balance;
+    const balancesByKey = await getUtxosBalances(utxosByKey);
+    const paperBalanceKeys = Object.keys(balancesByKey);
+    const hasBCH = paperBalanceKeys.includes("BCH");
 
-    // Get UTXOs associated with public address.
-    const u = await SLP.Address.utxo(fromAddr);
-    const utxos: UTXO[] = u.utxos;
+    // Generate a keypair from the WIF
+    const keyPair = SLP.ECPair.fromWIF(wif);
+    const fromAddr: string = SLP.ECPair.toCashAddress(keyPair);
 
     // Prepare to generate a transaction to sweep funds.
     const transactionBuilder = new SLP.TransactionBuilder(
       SLP.Address.detectAddressNetwork(fromAddr)
     );
-    let originalAmount: number = 0;
 
-    // Add all UTXOs to the transaction inputs.
-    for (let i = 0; i < utxos.length; i++) {
-      const utxo: utxo = utxos[i];
-      originalAmount = originalAmount + utxo.satoshis;
-      transactionBuilder.addInput(utxo.txid, utxo.vout);
-    }
+    if (tokenId && hasBCH) {
+      // The paper wallet has both BCH and SLP balances
+      // This case sweeps 1 SLP token and all of the BCH to the users wallet
+      // In the case the paper wallet has more than 1 SLP token, additional sweeps in the SLP only use case must be called
+
+      const scaledTokenSendAmount = new BigNumber(
+        balancesByKey[tokenId]
+      ).decimalPlaces(tokenDecimals);
+      const tokenSendAmount = scaledTokenSendAmount.times(10 ** tokenDecimals);
+      const sendOpReturn = slpjs.Slp.buildSendOpReturn({
+        tokenIdHex: tokenId,
+        outputQtyArray: [tokenSendAmount]
+      });
+      const tokenReceiverAddressArray = [addressSlp];
+
+      const slpUtxos = [...utxosByKey[tokenId]];
+      const bchUtxos = [...utxosByKey["BCH"]];
+
+      let inputUtxos = [...slpUtxos, ...bchUtxos];
+
+      let byteCount = SLPJS.calculateSendCost(
+        sendOpReturn.length,
+        inputUtxos.length,
+        tokenReceiverAddressArray.length + 1, // +1 to receive remaining BCH
+        fromAddr
+      );
```

Comment:
I love this. Is `SLPJS` the Simple Ledger SLP library? Or is this an internal library. I'll keep an eye on the tx fee while I test to make sure the fee is getting calculated correctly.

Reply:
Yeah, the simpleledger library
https://github.com/simpleledger/slpjs

The tx-fee here should be fine, the only known issue is with BIP70 SLP tx's with OP_RETURNS, which I'll fix next.
but a confirmation that the tx-fees are correct here would be great also.